src/main/java/com/cleanroommc/groovyscript/compat/mods/artisanworktables/Tables.java

```diff
@@ -0,0 +1,458 @@
+            msg.add(ingredients.isEmpty(), "Grid input must not be empty");
+            msg.add(ingredients.size() > width * height, () -> "maximum inputs are " + (width * height) + " but found " + ingredients.size());
+            if (msg.postIfNotEmpty()) return null;
+            ArtisanRecipeData recipe = new ArtisanRecipeData();
+            recipe.width = recipe.height = 0;
+            recipe.ingredients = ingredients;
+            return recipe;
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    @Property(property = "matrix", requirement = "groovyscript.wiki.craftingrecipe.matrix.required", virtual = true, defaultValue = "empty")
+    @Property(property = "key", defaultValue = "' ' = IIngredient.EMPTY", virtual = true)
+    @Property(property = "gridInput", valid = {@Comp(value = "1", type = Comp.Type.GTE), @Comp(value = "25", type = Comp.Type.LTE)}, virtual = true, defaultValue = "empty")
+    @Property(property = "input", valid = {@Comp(value = "1", type = Comp.Type.GTE), @Comp(value = "9", type = Comp.Type.LTE)}, value = "groovyscript.wiki.artisanworktables.tables.secondaryinputs.value")
+    public static class ArtisanRecipeBuilder extends AbstractRecipeBuilder<IArtisanRecipe> {
+        private final String registryName;
+        private final boolean shaped;
+        private final List<String> issues = new ArrayList<>();
+
+        @Property
+        private String name = "";
+
+        @Property
+        private boolean mirrored = false;
+
+        @Property(defaultValue = "empty")
+        private final Map<ResourceLocation, IRequirement> requirements = new HashMap<>();
+
+        @Property
+        private final List<OutputWeightPair> output = new ArrayList<>();
+
+        @Property(defaultValue = "empty", valid = @Comp(type = Comp.Type.LTE, value = "3"))
+        private final List<ToolEntry> tools = new ArrayList<>();
+
+        @Property(defaultValue = "true")
+        private boolean consumeSecondaryIngredients = true;
+
+        @Property
+        private int experience = 0;
+
+        @Property
+        private int level = 0;
+
+        @Property(defaultValue = "true")
+        private boolean consumeExperience = true;
+
+        @Property(defaultValue = "empty", valid = @Comp(type = Comp.Type.LTE, value = "3"))
+        private final List<ExtraOutputChancePair> extraOutputs = new ArrayList<>();
+
+        @Property(defaultValue = "the lowest tier that can hold all inputs of this recipe")
+        private int minimumTier = 0;
+
+        @Property(defaultValue = "2")
+        private int maximumTier = 2;
+
+        @Property
+        private boolean hidden = false;
+
+        private final AbstractCraftingRecipeBuilder<ArtisanRecipeData> gridBuilder;
+
+        private String makeArtisanName() {
+            return this.registryName + ':' + this.name;
+        }
+
+        public ArtisanRecipeBuilder(String name, boolean shaped) {
+            this.registryName = name;

```

Comment:
i dont see why this has to be set via the constructor - its only referenced in `validate()` and `register()`